rust/src/model.rs

```diff
@@ -0,0 +1,569 @@
+        let mut model = Model::<f64>::try_from(weights.clone()).unwrap();
+        assert_eq!(model.weights(), &weights);
+        assert_eq!(
+            Model::<f64>::try_from(inf_weights.clone()).unwrap_err(),
+            PetError::InvalidModel,
+        );
+        assert_eq!(
+            Model::<f64>::try_from(neg_inf_weights.clone()).unwrap_err(),
+            PetError::InvalidModel,
+        );
+        assert_eq!(
+            Model::<f64>::try_from(nan_weights.clone()).unwrap_err(),
+            PetError::InvalidModel,
+        );
+
+        // as ratio
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::<BigInt>::from_float(-1_f64).unwrap(),
+                Ratio::<BigInt>::zero(),
+                Ratio::<BigInt>::from_float(1_f64).unwrap(),
+            ],
+        );
+        model.weights = neg_inf_weights;
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::<BigInt>::from_float(f64::MIN).unwrap(),
+                Ratio::<BigInt>::zero(),
+                Ratio::<BigInt>::from_float(1_f64).unwrap(),
+            ],
+        );
+        model.weights = nan_weights;
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::<BigInt>::from_float(-1_f64).unwrap(),
+                Ratio::<BigInt>::zero(),
+                Ratio::<BigInt>::from_float(1_f64).unwrap(),
+            ],
+        );
+        model.weights = inf_weights;
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::<BigInt>::from_float(-1_f64).unwrap(),
+                Ratio::<BigInt>::zero(),
+                Ratio::<BigInt>::from_float(f64::MAX).unwrap(),
+            ],
+        );
+    }
+
+    #[test]
+    fn test_model_i32() {
+        // from
+        let weights = vec![-1_i32, 0_i32, 1_i32];
+        let model = Model::<i32>::from(weights.clone());
+        assert_eq!(model.weights(), &weights);
+
+        // as ratio
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::from_integer(BigInt::from(-1_i32)),
+                Ratio::<BigInt>::zero(),
+                Ratio::from_integer(BigInt::from(1_i32)),
+            ],
+        );
+    }
+
+    #[test]
+    fn test_model_i64() {
+        // from
+        let weights = vec![-1_i64, 0_i64, 1_i64];
+        let model = Model::<i64>::from(weights.clone());
+        assert_eq!(model.weights(), &weights);
+
+        // as ratio
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::from_integer(BigInt::from(-1_i64)),
+                Ratio::<BigInt>::zero(),
+                Ratio::from_integer(BigInt::from(1_i64)),
+            ],
+        );
+    }
+
+    // for float types the error depends on the order of magnitude of the weights => raise the
+    // tolerance or bound the random weights if this test fails to often
+    macro_rules! test_masking {
```

Comment:
Could we document what this test does? It seems important but I don't understand what it does.

Reply:
i'll do that. i still want to increase test coverage and will include it there.